src/bytes/bytes_lemmas.gobra

```diff
@@ -0,0 +1,359 @@
+}
+
+
+
+// Repeat
+
+ghost
+ensures bytes.Repeat(b, 1) == b
+decreases
+func lemmaRepeatOnce(b seq[byte]) {
+	assert bytes.Repeat(b, 1) == b ++ bytes.Repeat(b, 0)
+}
+
+ghost
+requires n > 0
+ensures bytes.Repeat(b, n) == bytes.Repeat(b, n-1) ++ b
+decreases n
+func lemmaRepeatFlip(b seq[byte], n int) {
+	if n == 1 {
+		assert bytes.Repeat(b, n-1) == seq[byte]{}
+	} else {
+		assert bytes.Repeat(b, n) == b ++ bytes.Repeat(b, n-1)
+		lemmaRepeatFlip(b, n-1)
+		assert bytes.Repeat(b, n-1) == bytes.Repeat(b, n-2) ++ b
+	}
+}
+
+ghost
+requires n >= 0
+ensures  b ++ bytes.Repeat(b, n) == bytes.Repeat(b, n) ++ b
+decreases
+func lemmaRepeatCommutative(b seq[byte], n int) {
+	assert b ++ bytes.Repeat(b, n) == bytes.Repeat(b, n+1)
+	lemmaRepeatFlip(b, n+1)
+}
+
+ghost
+requires n >= 0
+ensures  bytes.Repeat(b, 2 * n) == bytes.Repeat(b, n) ++ bytes.Repeat(b, n)
+decreases n
+func lemmaRepeat_2n(b seq[byte], n int) {
+	if n != 0 {
+		assert bytes.Repeat(b, n) == b ++ bytes.Repeat(b, n - 1)
+		assert bytes.Repeat(b, n) ++ bytes.Repeat(b, n) == b ++ bytes.Repeat(b, n - 1) ++ b ++ bytes.Repeat(b, n - 1)
+		// order of repeating doesn't matter because it's always the same element that is repeated
+		assert b ++ bytes.Repeat(b, n - 1) ++ b ++ bytes.Repeat(b, n - 1) == b ++ (bytes.Repeat(b, n - 1) ++ b) ++ bytes.Repeat(b, n - 1)
+		lemmaRepeatCommutative(b, n - 1)
+		assert b ++ (bytes.Repeat(b, n - 1) ++ b) ++ bytes.Repeat(b, n - 1) == b ++ (b ++ bytes.Repeat(b, n - 1)) ++ bytes.Repeat(b, n - 1)
+		assert b ++ bytes.Repeat(b, n - 1) ++ b ++ bytes.Repeat(b, n - 1) == b ++ b ++ bytes.Repeat(b, n - 1) ++ bytes.Repeat(b, n - 1)
+		lemmaRepeat_2n(b, n - 1)
+		assert bytes.Repeat(b, n - 1) ++ bytes.Repeat(b, n - 1) == bytes.Repeat(b, 2 * (n - 1))
+		assert bytes.Repeat(b, 2 * n - 1) == b ++ bytes.Repeat(b, n - 1) ++ bytes.Repeat(b, n - 1)
+		assert bytes.Repeat(b, 2 * n) == b ++ b ++ bytes.Repeat(b, n - 1) ++ bytes.Repeat(b, n - 1)
+	} else {
+		assert bytes.Repeat(b, 2 * 0) == bytes.Repeat(b, 0) ++ bytes.Repeat(b, 0)
+	}
+}
+
+ghost
+requires bp < lnb
+requires bp == lb * i
+requires lnb == lb * count
+requires 0 <= lb
+ensures  i < count
+decreases
+func lemmaStillHaveToRepeat(i, count, bp, lb, lnb int) {
+	if lb != 0 {
+		assert lb * i < lnb
+		assert lb * i < lb * count
+		LemmaMulInjLtLeftInv(i, count, lb)
+		assert i < count
+	}
+}
+
+
+
+
+
+ghost
+requires a == b
+requires count >= 0
+requires a >= 0
+requires nb[:MinInt(count, a) * len(s)] == bytes.Repeat(s, MinInt(count, a))
+ensures  nb[:MinInt(count, b) * len(s)] == bytes.Repeat(s, MinInt(count, b))
+decreases
+func lemmaSubstV(nb, s seq[byte], count, a, b int ) { }
+
+ghost
+requires InRangeInc(bp, 0, len(nb))
+requires InRangeInc(nCopied, 0, bp)
+requires bp + nCopied <= len(nb)
+requires acc(sl.Bytes(nb, 0, len(nb)), R50)
+requires unfolding acc(sl.Bytes(nb, 0, len(nb)), R50) in forall i int :: {&nb[bp:][i]} 0 <= i && i < len(nb[bp:]) ==> &nb[bp:][i] == &nb[i+bp]
```

Comment:
why do we need this precondition here? we should be able to establish this in the body, as this is always true. It seems unnecessary to put a tautology in a precondition

Reply:
In the following precondition, we talk about nb[bp:][i], which gobra does not automatically understand is the same as nb[bp+i]. This precondition is always trivially satisfied but is needed for the specification to be well-formed

Reply:
this might not be needed if we have use a pure function instead